restore/add F11 to toggle fullscreen

ronso0 opened the pull request:
... in addition to the system default set by Qt according to the OS / desktop environment used. https://github.com/mixxxdj/mixxx/pull/2735#issuecomment-623171144

1) to restore previous Linux shortcut
2) to have it as additional shortcut on OS where QKeySequence::FullScreen is not set (for example Ubuntu Studio 20.04 as of 2020-05-04)

ronso0 commented:
appearantly the SVG didn't get updated due to some text/script error.
Will fix..

daschuer reviewed (changes requested):
Thank you for taking care. I was tempted to fix it as well, but I did not do It because we don't follow the QT standard. Please add a comment why that is OK.

Holzhaus commented:
Can we have F11 *in addition* to the `QKeySequence::Fullsceeen` or is this not possible?

ronso0 commented:
> Can we have F11 _in addition_ to the `QKeySequence::Fullsceen` or is this not possible?

Should be possible according to https://doc.qt.io/qt-5/qaction.html#setShortcuts
Seems it wouldn't work on MacOS though.

Holzhaus commented:
If we use `#ifdef __LINUX__` that should not be a problem.

ronso0 commented:
Alright, what do others think?
I don't mind having both as long as we have the good ol' F11 back

daschuer commented:
If both works OK else I would vote for F11. How does it look in the manu bar with both?

ronso0 commented:
According to the docs, the first key sequence is picked for the menu.
I'll try it...

ronso0 commented:
Alright, with the last commit the View menu looks like before (Fullscreen Ctrl + F11), but F11 works, too.
Question is: what do we need Ctrl + F11 for?
* Do we want to keep it for users that use mouse gestures or other system meta shortcuts that are directed to Ctrl + F11 and don't look into the menu? Then, traditional F11 is not discoverable.
* The other way around: have F11 listed in the menu, and have Ctrl + F11 as addiition.

ronso0 committed:
use F11 as additional fullscreen shortcut

1) to restore previous Linux shortcut
2) to have it as additional shortcut on OS where
   QKeySequence::FullScreen is not set (for example Ubuntu
   Studio 20.04 as of 2020-05-04)

daschuer commented:
It is fine now. The shortcut QT thinks is right is listed. No one can blame us, but users trying F11 because they are used to are not disappointed.

Please add a brief comment about that to the code.

esbrandt commented:
It reverts the previous change to the full screen menu for Linux that happened some month ago in #2178. The intention was to have a standard shortcut, depending on the Operating system. It got unnoticed, that on linux the shortcut is also determined by the desktop environment used.

So since there was never a release with the changed shortcut, any reason to allow more shortcuts than the established ``F11``? This is PR basically corrects a faulty change.

daschuer commented:
The key is hard coded here:  
https://github.com/qt/qtbase/blob/68a20d69419671f28990052aca4c89a52ee78046/src/gui/kernel/qplatformtheme.cpp#L341

```
    {QKeySequence::FullScreen,              1,          Qt::META | Qt::CTRL | Qt::Key_F,        KB_Mac},
    {QKeySequence::FullScreen,              0,          Qt::ALT  | Qt::Key_Enter,               KB_Win},
    {QKeySequence::FullScreen,              0,          Qt::CTRL | Qt::SHIFT | Qt::Key_F,       KB_KDE},
    {QKeySequence::FullScreen,              1,          Qt::CTRL | Qt::Key_F11,                 KB_Gnome},
    {QKeySequence::FullScreen,              1,          Qt::Key_F11,                            KB_Win | KB_KDE},
```

daschuer commented:
Scary ... 
maybe we should remove the conditional and add always F11

ronso0 commented:
Anyway, we need to document F11 in the keyboard section of the manual, otherwise users might think all F keys are free for manual mapping.

esbrandt commented:
Still listed with ``F11``


https://github.com/mixxxdj/manual/blob/manual-2.2.x/source/chapters/appendix.rst#application-shortcuts

ronso0 commented:
Okay, so we need to update the manual anyway.

FWIW In mixxx I just built in Ubuntu Studio 20.04 there's neither a shortcut printed in the menu, nor does Ctrl+F11 or F11 work, just clicking Fullscreen in the menu does.
This is with Xfce 4.14

ronso0 commented:
> Scary ...
> maybe we should remove the conditional and add always F11

Jup, I think so, too, now. Ubuntu (Studio) 20.04 is out now and there I don't get a hotkey at all.

ronso0 commented:
I removed the #ifdef __LINUX__

daschuer commented:
Thank you. Still working :-)

nopeppermint commented:
this brakes for me the F11 toggle full-screen on linux, as described here : https://bugs.launchpad.net/mixxx/+bug/1882474
without those changes, it's working !

uklotzde commented:
@nopeppermint I cannot confirm this. Toggling full-screen with F11 works fine on GNOME.

nopeppermint commented:
I am using Linux Mint with Cinnamon..

ronso0 commented:
what's the regular fullscreen hotkey in Cinnamon? (for other apps, probably listed in system settings > keyboard) 
I suspect that something weird is going on in case F11 is the system fullscreen shortcut and we add F11 again..

nopeppermint commented:
F11 in Firefox

Holzhaus commented:
Could this be an upstream bug in Cinnamon? I'd expect that the system fullscreen shortcut is not passed through to client applications if it's already handles by the window manager.

ronso0 commented:
@nopeppermint Could you please build https://github.com/ronso0/mixxx/tree/fullscreen-debug and paste the output here?
It's printed very early when the launch image is shown. For me it prints this (OS fullscreen key not set):
```
Debug [Main]: 
Debug [Main]:        OS fullscreen key: QKeySequence::FullScreen
Debug [Main]:           keys list:
Debug [Main]:                 QKeySequence("")
Debug [Main]:                 QKeySequence("F11")
Debug [Main]: 
```

btw I very much recommend building with `cmake`.

ronso0 commented:
> I try if we can filter that out and then show F11 only.

okay, the follow-up PR is https://github.com/mixxxdj/mixxx/pull/3011

nopeppermint commented:
I just build your branch (with scons, as cmake is not yet running on my system) and got the following output, hope this helps:
```
 ./mixxx --developer
Debug [Main]: using qt5ct plugin
Info [Main]: faad2::LibLoader - Successfully loaded FAAD2 library "libfaad.so" version 2.8.8
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "aif"
Info [Main]: SoundSourceProxy -   2 : "libsndfile"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "aiff"
Info [Main]: SoundSourceProxy -   2 : "libsndfile"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "caf"
Info [Main]: SoundSourceProxy -   1 : "libsndfile"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "flac"
Info [Main]: SoundSourceProxy -   3 : "Xiph.org libFLAC"
Info [Main]: SoundSourceProxy -   1 : "libsndfile"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "m4a"
Info [Main]: SoundSourceProxy -   2 : "Nero FAAD2"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "mp3"
Info [Main]: SoundSourceProxy -   2 : "MAD: MPEG Audio Decoder"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "mp4"
Info [Main]: SoundSourceProxy -   2 : "Nero FAAD2"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "ogg"
Info [Main]: SoundSourceProxy -   3 : "Xiph.org OggVorbis"
Info [Main]: SoundSourceProxy -   1 : "libsndfile"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "opus"
Info [Main]: SoundSourceProxy -   3 : "Xiph.org libopusfile"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "wav"
Info [Main]: SoundSourceProxy -   2 : "libsndfile"
Debug [Main]: "Mixxx" "2.3.0-beta" "(git fullscreen-debug r7528; built on: Aug 11 2020 @ 18:03:48; flags: asan=0 battery=1 buildtime=1 bulk=1 color=0 coreaudio=0 faad=1 ffmpeg=0 hid=1 hss1394=0 lilv=1 localecompare=1 macappstore=0 mad=1 mediafoundation=0 modplug=0 optimize=portable opus=1 perftools=0 perftools_profiler=0 profiling=0 qt_sqlite_plugin=0 qtkeychain=0 shoutcast=1 test=False tsan=0 ubsan=0 verbose=1 vinylcontrol=1 wv=0)" is starting...
Debug [Main]: Compile time library versions:
Debug [Main]: Qt: 5.9.5
Debug [Main]: libshout: 2.4.1
Debug [Main]: PortAudio: 1246720 PortAudio V19.6.0-devel, revision 396fe4b6699ae929d3a685b3ef8a7e97396139a4
Debug [Main]: RubberBand: 1.8.1
Debug [Main]: SoundTouch: 2.1.1
Debug [Main]: TagLib: 1.11.1
Debug [Main]: ChromaPrint: 1.4.3
Debug [Main]: Vorbis: Xiph.Org libVorbis 1.3.5
Debug [Main]: libsndfile: libsndfile-1.0.28
Debug [Main]: FLAC: 1.3.2
Debug [Main]: libmp3lame: 3.100
Debug [Main]: QStandardPaths::writableLocation(HomeLocation): "/home/stefan"
Debug [Main]: QStandardPaths::writableLocation(DataLocation): "/home/stefan/.local/share/Mixxx"
Debug [Main]: QCoreApplication::applicationDirPath() "/home/stefan/mixxx_neu/mixxx"
Debug [Main]: Loading resources from  "/home/stefan/mixxx_neu/mixxx/res/"
Debug [StatsManager]: StatsManager thread starting up.
Debug [Main]: Configuration file is at the current version 2.3.0-beta
Info [Main]: BroadcastSettings - Found 1 profile(s)
Debug [Main]: Loading resources from  "/home/stefan/mixxx_neu/mixxx/res/"
Debug [Main]: Found and will use default keyboard preset "/home/stefan/mixxx_neu/mixxx/res/keyboard/de_CH.kbd.cfg"
Debug [Main]: Loading resources from  "/home/stefan/mixxx_neu/mixxx/res/"
Debug [Main]: Loading translations for locale "de_CH" from translations folder "/home/stefan/mixxx_neu/mixxx/res/translations/" : success
Debug [Main]: D-Bus global menu: no
Debug [Main]: 
Debug [Main]:        OS fullscreen key: QKeySequence::StandardKey(FullScreen)
Debug [Main]:           keys list:
Debug [Main]:                 QKeySequence("F11")
Debug [Main]:                 QKeySequence("F11")
Debug [Main]: 
Debug [Main]: Loading resources from  "/home/stefan/mixxx_neu/mixxx/res/"
Debug [Main]: FONT LOADED family: "Open Sans" style: "Bold" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "Bold" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
```

ronso0 commented:
alright, so it's as I expected, you have F11 twice in the shortcut list.
I try to filter that out, as well.
..continued in #3011
